### PR TITLE
Simple render

### PR DESCRIPTION
**Pull request by Brookoko**

### CODE REVIEW

File: src/main/java/labs/introtoprogramming/lab5/exception/NoCamerException.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package labs.introtoprogramming.lab5.exception;
+
+public class NoCamerException extends RuntimeException {
```
Comment (nikitavbv) [inline src/main/java/labs/introtoprogramming/lab5/exception/NoCamerException.java]: `NoCameraexception`

File: src/main/java/labs/introtoprogramming/lab5/geometry/Ray.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package labs.introtoprogramming.lab5.geometry;
+
+public class Ray {
+  private Vector3 origin;
+  private Vector3 direction;
+  private double t;
```
Comment (nikitavbv) [inline src/main/java/labs/introtoprogramming/lab5/geometry/Ray.java]: Add a comment explaining what t is or choose more descriptive name

File: src/main/java/labs/introtoprogramming/lab5/geometry/Vector3.java [hunk 3]
```diff
@@ -11,4 +11,88 @@ public Vector3(double x, double y, double z) {
     this.y = y;
     this.z = z;
   }
+
+  public double magnitude() {
+    return Math.sqrt(sqrtMagnitude());
+  }
+
+  public double sqrtMagnitude() {
```
Comment (nikitavbv) [inline src/main/java/labs/introtoprogramming/lab5/geometry/Vector3.java]: `magnitudeSquared` is more descriptive

File: src/main/java/labs/introtoprogramming/lab5/geometry/Vector3.java [hunk 4]
```diff
@@ -11,4 +11,88 @@ public Vector3(double x, double y, double z) {
...
+    double magnitude = magnitude();
+    if (magnitude > 0) {
+      double inverseMagnitude = 1 / magnitude;
+      return multiply(inverseMagnitude);
+    }
+    return multiply(1);
+  }
+
+  public double distance(Vector3 v) {
+    return new Vector3(
```
Comment (nikitavbv) [inline src/main/java/labs/introtoprogramming/lab5/geometry/Vector3.java]: Maybe replace this with `subtract`?

File: src/main/java/labs/introtoprogramming/lab5/geometry/Vector3.java [hunk 5]
```diff
@@ -11,4 +11,88 @@ public Vector3(double x, double y, double z) {
...
+  public Vector3 add(Vector3 v) {
+    return new Vector3(x + v.x, y + v.y, z + v.z);
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (!(obj instanceof Vector3)) return false;
+    if (obj == this) return true;
+    Vector3 vector = (Vector3) obj;
+    return Math.abs(vector.x - x) < 1e-10 &&
```
Comment (nikitavbv) [inline src/main/java/labs/introtoprogramming/lab5/geometry/Vector3.java]: Make 1e10 a constant, something like `COORDINATES_PRECISION` (with a short comment explaining)

File: src/main/java/labs/introtoprogramming/lab5/geometry/Vector3.java [hunk 6]
```diff
@@ -11,4 +11,88 @@ public Vector3(double x, double y, double z) {
...
+            Math.abs(vector.z - z) < 1e-10;
+  }
+
+  @Override
+  public int hashCode() {
+    return (int) (x + y + z);
+  }
+
+  public static Vector3 zero() {
+    return new Vector3(0, 0, 0);
```
Comment (nikitavbv) [inline src/main/java/labs/introtoprogramming/lab5/geometry/Vector3.java]: Maybe make this a global static variable instead (`ZERO`) - similar to `BigInteger.ONE`

File: src/main/java/labs/introtoprogramming/lab5/object/Sphere.java [hunk 7]
```diff
@@ -1,11 +1,45 @@
...
 import labs.introtoprogramming.lab5.geometry.Vector3;
 import labs.introtoprogramming.lab5.scene.SceneObject;
+import labs.introtoprogramming.lab5.scene.Transform;
 
 public class Sphere extends SceneObject {
 
-  private Vector3 position;
   private double radius;
 
+  public Sphere(Transform transform, double radius) {
```
Comment (nikitavbv) [inline src/main/java/labs/introtoprogramming/lab5/object/Sphere.java:18]: Add a short javadoc here, explaining what is transform and why it is needed.

File: src/main/java/labs/introtoprogramming/lab5/scene/Camera.java [hunk 8]
```diff
@@ -1,4 +1,22 @@
 package labs.introtoprogramming.lab5.scene;
 
 public abstract class Camera extends SceneObject {
+  private double size;
+
+  public Camera(Transform transform, double size) {
```
Comment (nikitavbv) [inline src/main/java/labs/introtoprogramming/lab5/scene/Camera.java]: Add a javadoc here explaining what `size` is.

File: src/main/java/labs/introtoprogramming/lab5/scene/Camera.java [hunk 9]
```diff
@@ -1,4 +1,22 @@
...
+    super(transform);
+    this.size = size;
+  }
+
+  public Camera(Transform transform) {
+    this(transform, 1);
```
Comment (nikitavbv) [inline src/main/java/labs/introtoprogramming/lab5/scene/Camera.java]: Move 1 to a constant. Like `DEFAULT_SIZE`

File: src/main/java/labs/introtoprogramming/lab5/scene/Camera.java [hunk 10]
```diff
@@ -1,4 +1,22 @@
...
+  }
+
+  public double size() {
+    return size;
+  }
+
+  public double fieldOfView() {
+    return 2 * Math.atan(size);
```
Comment (nikitavbv) [inline src/main/java/labs/introtoprogramming/lab5/scene/Camera.java:40]: Short explanation in a comment here would be nice :)

File: src/main/java/labs/introtoprogramming/lab5/scene/RasterRender.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  @Override
+  public void render(Scene scene) {
+    Camera camera = scene.getCamera().orElseThrow(NoCamerException::new);
+    intractable = scene.getSceneObjects().stream()
+            .filter(obj -> !(obj instanceof PointLight) && !(obj instanceof Camera))
+            .collect(Collectors.toList());
+    Vector3 pos = camera.getTransform().position();
+    double aspectRation = (double) width / height;
```
Comment (nikitavbv) [inline src/main/java/labs/introtoprogramming/lab5/scene/RasterRender.java]: Maybe move calculation of aspectRatio and fieldOfView to camera class?

File: src/main/java/labs/introtoprogramming/lab5/scene/SceneObject.java [hunk 12]
```diff
@@ -1,9 +1,34 @@
...
-import labs.introtoprogramming.lab5.geometry.Vector3;
+import labs.introtoprogramming.lab5.geometry.Ray;
 
 public class SceneObject {
 
-  private Vector3 position;
+  protected Transform transform;
+  protected Mesh mesh;
 
+  public SceneObject(Transform transform, Mesh mesh) {
```
Comment (nikitavbv) [inline src/main/java/labs/introtoprogramming/lab5/scene/SceneObject.java:16]: Add a short javadoc explaining why transform is needed.

File: src/main/java/labs/introtoprogramming/lab5/scene/Transform.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+  public Transform(Vector3 position) {
+    this(position, Vector3.zero());
+  }
+
+  public Transform() {
+    this(Vector3.zero(), Vector3.zero());
+  }
+
+  public Vector3 position() {
+    return position.multiply(1);
```
Comment (nikitavbv) [inline src/main/java/labs/introtoprogramming/lab5/scene/Transform.java]: Why not return position? Or just clone it?

File: src/test/java/labs/introtoprogramming/lab5/geometry/RayTests.java [hunk 14]
```diff
@@ -0,0 +1,31 @@
+package labs.introtoprogramming.lab5.geometry;
+
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+
+public class RayTests {
+  private static final double DELTA = 1e-10;
```
Comment (nikitavbv) [inline src/test/java/labs/introtoprogramming/lab5/geometry/RayTests.java:8]: Nice!

File: src/test/java/labs/introtoprogramming/lab5/geometry/VectorTests.java [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+  @Test
+  public void testMagnitude() {
+    Vector3 vector = Vector3.right();
+    assertEquals(1, vector.magnitude(), DELTA);
+  }
+
+  @Test
+  public void testNormalize() {
+    Vector3 a = new Vector3(2, 4, 4);
+    assertEquals(new Vector3( 2. / 6, 4. / 6, 4. / 6), a.normalize());
```
Comment (nikitavbv) [inline src/test/java/labs/introtoprogramming/lab5/geometry/VectorTests.java]: `2.0` please. It is more readable.

File: src/main/java/labs/introtoprogramming/lab5/scene/RasterRender.java (hunk 11 above)
Comment (Brookoko) [reply src/main/java/labs/introtoprogramming/lab5/scene/RasterRender.java]: AspectRation is defined by image width and height not by camera linear parameters, so I think it should not be in camera class.

File: src/main/java/labs/introtoprogramming/lab5/scene/Transform.java (hunk 13 above)
Comment (Brookoko) [reply src/main/java/labs/introtoprogramming/lab5/scene/Transform.java]: I create copy of vector by multiplying it by one or is there better way?

Comment (nikitavbv) [reply src/main/java/labs/introtoprogramming/lab5/scene/Transform.java]: Better implement Cloneable and use `clone()`

**Pull request merged**